src/util.ts

```diff
@@ -46,19 +50,42 @@ export const sumByRowcount = (num: number) =>
   _.sumBy(({ rowcount }) => (rowcount === num ? 1 : 0))
 
 /**
- * Get the documents that failed to be written during a bulk
+ * Get the document ids that failed to be written during a bulk
  * query.
  */
 export const getFailedRecords = (
   results: BulkQueryResult['results'],
-  documents: Document[]
+  documents: ChangeStreamInsertDocument[]
 ) => {
   const failed: unknown[] = []
   for (let i = 0; i < results.length; i++) {
     const result = results[i]
     if (result.rowcount === -2) {
-      failed.push(documents[i])
+      failed.push(documents[i].documentKey._id)
     }
   }
   return failed
 }
+
+export const partitionEvents = (docs: ChangeStreamDocument[]) => {
+  const groups = []
+  let subGroup = []
+  let previousOperationType: ChangeStreamDocument['operationType'] | undefined =
```

Comment:
We could simplify this code and make the bulk inserts more efficient (only one bulk insert per batch), by doing something like this:

```typescript
const inserts = docs.filter((doc) => doc.operationType === 'insert')
const nonInserts = docs.filter((doc) => doc.operationType !== 'insert')

return { inserts, nonInserts }
```

Then you could do a single bulk insert with the `inserts`, and then process the `nonInserts` (updates/deletes) one at a time.

Is it safe to reorder the operations like this, so that the inserts all happen at the beginning? I think the answer is "yes".

Consider the following stream of changes:

```
Insert Record 1
Insert Record 2
Insert Record 3
Update Record 2
Delete Record 1
Insert Record 4
Update Record 4
Update Record 4
Insert Record 5
Insert Record 6
```

Note that it's safe to reorder it so that all of the inserts happen first as a single batch:


```
Batch Insert:
  Insert Record 1
  Insert Record 2
  Insert Record 3
  Insert Record 4
  Insert Record 5
  Insert Record 6

(single operations)
Update Record 2
Delete Record 1
Update Record 4
Update Record 4
```

This is because inserts have "no dependencies". Those updates and deletes (of other records) that happened between the inserts had no impact at all on the ability to do the inserts. Reordering in this manner is therefore a safe thing to do, and it could be worthwhile, as the code would be simpler, and we would be doing more of the database operations in bulk.

The batch failure/retry scenario seems like it should be OK, too.

Reply:
Discussed this idea after standup, and there is one problem: you can delete a Mongo record, and then create a new record with the same `_id`. If we reordered the operations like above to put all the inserts first, that would result in the wrong result -- the record would not exist, when it should.